File: sdk/ruby/lib/anyvali/schemas/enum_schema.rb
# frozen_string_literal: true

module AnyVali
  class EnumSchema < Schema
    attr_reader :values

    def initialize(values:, **kwargs)
      @values = values.freeze
      super(kind: "enum", **kwargs)
    end

    def to_node
      node = super
      node["values"] = @values
      node
    end

    protected

    def validate(value, path, issues, context)
      match = @values.any? { |v| v.class == value.class && v == value }
      # Special nil handling
      match ||= @values.include?(nil) && value.nil?

      unless match
        enum_str = @values.map(&:to_s).join(",")
        received_str = value.nil? ? "null" : value.to_s
        issues << ValidationIssue.new(
          code: IssueCodes::INVALID_TYPE,
          path: path,
          expected: "enum(#{enum_str})",
          received: received_str
        )
      end
    end

    def dup_with(**overrides)
      attrs = {
        values: @values,
        kind: @kind,
        constraints: @constraints,
        coerce_config: @coerce_config,
        default_value: @default_value,
        has_default: @has_default,
        custom_validators: @custom_validators
      }.merge(overrides)
      self.class.new(**attrs)
    end
  end
end

File: sdk/csharp/src/AnyVali/Schemas/EnumSchema.cs
namespace AnyVali.Schemas;

public sealed class EnumSchema : Schema<object?>
{
    private readonly List<object> _values;

    public EnumSchema(IEnumerable<object> values)
    {
        _values = values.ToList();
    }

    internal override object? Validate(object? input, ValidationContext ctx)
    {
        foreach (var v in _values)
        {
            if (v is null && input is null) return input;
            if (v is not null && input is not null)
            {
                // Numeric comparison
                if (IsFiniteNumber(v) && IsFiniteNumber(input) && ToDouble(v) == ToDouble(input))
                    return input;
                if (v.Equals(input))
                    return input;
            }
        }

        var valuesStr = string.Join(",", _values.Select(v => v?.ToString() ?? "null"));
        ctx.Issues.Add(new ValidationIssue
        {
            Code = IssueCodes.InvalidType,
            Message = $"Expected one of enum({valuesStr}), received {input}",
            Path = ctx.ClonePath(),
            Expected = $"enum({valuesStr})",
            Received = input?.ToString() ?? "null",
        });
        return null;
    }

    internal override Dictionary<string, object?> ToNode()
    {
        var node = new Dictionary<string, object?>
        {
            ["kind"] = "enum",
            ["values"] = _values.Select(v => (object?)v).ToList(),
        };
        AddDefaultAndCoercion(node);
        return node;
    }

    internal override Schema Clone() => new EnumSchema(_values)
    {
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable,
    };
}
